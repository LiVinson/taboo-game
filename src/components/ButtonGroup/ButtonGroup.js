import React from "react"
import PropTypes from "prop-types"

//Appears at bottom of Taboo Card when a buttons prop is passed to taboo.
// Receives array of objects with button text and click handler
export default function ButtonGroup({ buttons }) {
  const { handleClick: leftHandleClick, text: leftText } = buttons[0]
  const { handleClick: rightHandleClick, text: rightText } = buttons[1]

  return (
    <div>
      <button onClick={leftHandleClick}>{leftText}</button>
      <button onClick={rightHandleClick}>{rightText}</button>
    </div>
  )
}

ButtonGroup.propTypes = {
  buttons: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      handleClick: PropTypes.func.IsRequired,
    })
  ),
}
