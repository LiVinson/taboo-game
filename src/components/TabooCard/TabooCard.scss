.taboo-card {
    width:30rem; 
    margin: 0 auto;
    padding: 2rem;
    border-radius: 2px;
    position: relative;
    z-index: 15;
    background-color: var(--color-primary);

    box-shadow: 0px .8rem 1rem rgba(0,0,0,.9);
    //Used for inheritance on child components
    color: var(--color-primary-light);
    


    //Main taboo card on home screen, make text and padding larger
    &--home &__word-container &__word {
        font-size: 4rem;
    }

    &--home &__list-container {
        padding-top:2rem;
        padding-bottom:2rem;
    }

    &--bottom {    
        color:  var( --color-tertiary);
        background-color: currentColor;
        position: absolute;
        bottom:9rem;
        right: 8rem;
        z-index: 5;
        transform: rotate(-42deg);
    }

    &--middle {
        color:  var( --color-secondary);
        background-color: currentColor;
        position: absolute;
        bottom:8rem;
        left:7rem;
        z-index: 10;
        transform: rotate(42deg)
    }

    &__word-container
    {
        background-color: var(--color-off-white);
        padding: 1.2rem;
        text-align: center;   
        margin-bottom: 1.5rem;
        border-radius: 2px;
        color:inherit;

    }

    &__word {
        font-family: 'Oswald', sans-serif;
        font-size: 3.5rem;
        color: inherit;
        font-weight: 600;
        text-transform: uppercase;
    
    }

    &__list-container {
        background-color: var(--color-off-white);
        padding: 1.5rem;
        text-align: center;  
        border-radius: 2px;
    }

    &__list {
        list-style-type: none;

    }

    &__list-item {
        font-family: 'Oswald', sans-serif;
        font-size: 2.5rem;
        position: relative;
        padding-bottom:1.5rem;
    }


    &__list-item:after {
        content:"";
        display:block;
        width: 10rem;
        margin: 0 auto;
        padding-top:1rem;
        border-bottom: #C4C4C4 2px solid; 
    }
}