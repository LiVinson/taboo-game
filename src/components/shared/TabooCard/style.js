import styled, { css } from 'styled-components'

//Primary Taboo Card container
export const StyledTabooCard = styled.div`
	width: 30rem;
	margin: 0 auto;
	margin-bottom: ${(props) => props.margin && '3rem'};
	padding: 2rem;
	border-radius: 2px;
	position: relative;
	background-color: currentColor;
	color: ${(props) => props.theme.color.primary};

	/* Creates a box shadow on bottom left corner of card  */
	&:before {
		z-index: -1;
		position: absolute;
		content: '';
		bottom: 1.2rem;
		left: 0.5rem;
		width: 50%;
		top: 80%;
		max-width: 300px;
		background: #777;
		box-shadow: 0 1.5rem 1rem rgba(0, 0, 0, 0.4);
		transform: rotate(-3deg);
	}
`

StyledTabooCard.displayName = 'StyledTabooCard'

//Creates the "stacked" background element. Color and z-index varies based on level in stack.
export const StyledLayeredTabooCard = styled(StyledTabooCard)`
	position: absolute;
	top: 66%;
	left: 50%;
	margin-bottom: 0rem;
	${({ layer }) =>
		layer === 'top' &&
		css`
			color: ${(props) => props.theme.color.accent2};
			z-index: -5;
			transform: translate(-50%, -50%) rotate(-20deg);
		`}

		${({ layer }) =>
			layer === 'middle' &&
			css`
				color: ${(props) => props.theme.color.accent1};
				z-index: -10;
				transform: translate(-50%, -50%) rotate(16deg);
			`}

		${({ layer }) =>
			layer === 'bottom' &&
			css`
				color: ${(props) => props.theme.color.accent3};
				z-index: -15;
				transform: translate(-50%, -50%) rotate(-5deg);
			`}
	/* Removes the box shadow from left corner from standard taboo card */
	&:before {
		display: none;
	}
	/* Creates an overlay on the card so the colors are muted and appears more in the background */
	&:after {
		display: inline-block;
		content: '';
		position: absolute;
		top: 0;
		left: 0;
		width: 100%;
		height: 100%;
		background-color: lightgray;
		opacity: 0.5;
	}
 `

StyledLayeredTabooCard.displayName = 'StyledLayeredTabooCard'

// Used for top portion of taboo card that contains word to guess
export const TabooWordContainer = styled.div`
	background-color: ${(props) => props.theme.color.offWhite};
	padding: 1.2rem;
	text-align: center;
	border-radius: 2px;
	font-family: ${(props) => props.theme.text.display};
	font-size: ${(props) => props.theme.fontSize.large};
	color: inherit;
	font-weight: 400;
	text-transform: uppercase;
`
TabooWordContainer.displayName = 'TabooWordContainer'

//Bottom portion of Taboo card. Contains list of taboo words or whatever was passed in as children
export const TabooBody = styled.div`
	background-color: ${(props) => props.theme.color.offWhite};
	padding: 1.5rem;
	text-align: center;
	border-radius: 2px;
	font-size: 4rem;
	margin-top: 1.5rem;
`
TabooBody.displayName = 'TabooBody'

export const StyledListItem = styled.li`
	font-family: ${(props) => props.theme.text.display};
	font-size: ${(props) => props.theme.fontSize.medium};
	position: relative;
	padding-bottom: 1.5rem;
`

StyledListItem.displayName = 'StyledListItem'

export const StyledListItemSecondary = styled.li`
	color: ${(props) => props.theme.color.grayDark2};
	font-size: 2.2rem;
`
export const StyledListTitle = styled.p`
	text-transform: uppercase;
	color: ${(props) => props.theme.color.grayDark2};
	font-size: 2.5rem;
	margin-bottom: 0.3rem;
	font-family: ${(props) => props.theme.text.display};

	/* Faint gray line after section title*/
	&::after {
		display: block;
		content: '';
		width: 80%;
		height: 1.5px;
		background-color: ${(props) => props.theme.color.lightGray2};
		margin: 0.8rem auto 0.6rem auto;
	}

	
	&:not(:first-child) {
    margin-top:1rem;
  }
`
