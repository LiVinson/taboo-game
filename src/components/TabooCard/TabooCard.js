import React from "react"
import "./TabooCard.scss"
import ButtonGroup from "components/ButtonGroup"
import PropTypes from "prop-types"
/*
Props:
   type(Used for styling cards differently)
  tabooWord: key What is at top of card
  list: Array of words to display in card body 
   */

export default function TabooCard({
  type,
  tabooWord,
  list,
  buttons,
  children,
    
}) {
  let secondaryClass = null
  //types: home, middle, bottom
  if (type) {
    secondaryClass = `taboo-card--${type}`
  }  
  return (
    //Wraps entire card
    <div className={`taboo-card ${secondaryClass}`}>
      {/* Contains taboo word*/}
      <div className="taboo-card__word-container">
        <p className="taboo-card__word">{tabooWord}</p>
      </div>
      {/* If there are children, render them as is. Otherwise, loop over list of items and display */}
      <div className="taboo-card__list-container">
        {children ? (
          children
        ) : (
          <ul className="taboo-card__list">
            {list.map((item, index) => (
              <li key={index} className="taboo-card__list-item">
                {item}
              </li>
            ))}
          </ul>
        )}
        { buttons ? <ButtonGroup buttons={buttons} /> : null}
      </div>
    </div>
  )
}

TabooCard.propTypes = {
  type: PropTypes.string,
  tabooWord: PropTypes.string.isRequired,
  list: PropTypes.array,
  buttons: PropTypes.arrayOf(PropTypes.shape({
    type: PropTypes.string,
    click: PropTypes.func
  }))
}