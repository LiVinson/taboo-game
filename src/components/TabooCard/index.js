import TabooCard, { TabooCardTop } from "./TabooCard"

export { TabooCardTop };
export default TabooCard;
