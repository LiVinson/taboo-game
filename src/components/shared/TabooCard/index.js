		//If item is an object, filter based on property  value. Otherwise filter based on value of item (i.e. string)
import { TabooCard, ButtonTabooCard, LayeredTabooCard, TabooCardTop, FilteredTabooList} from "./TabooCard"

		//If item is an object, filter based on property  value. Otherwise filter based on value of item (i.e. string)
export { TabooCard, ButtonTabooCard,  LayeredTabooCard, TabooCardTop, FilteredTabooList};

