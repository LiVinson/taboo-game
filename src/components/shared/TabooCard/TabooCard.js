import React from 'react'
import PropTypes from 'prop-types'
import {
	StyledTabooCard,
	StyledLayeredTabooCard,
	TabooWordContainer,
	TabooBody,
	StyledListItem,
	StyledListItemSecondary,
	StyledListTitle,
} from './style'
import List from 'components/shared/List'
import ButtonContainer from 'components/ButtonContainer'
import Star from 'components/shared/Star'
import {ErrorMessage} from 'components/shared/FeedbackMessage'

//Primary TabooCard component used for main menu and game
export const TabooCard = ({ tabooWord, list, children }) => {
	return (
		<StyledTabooCard>
			<TabooWordContainer>{tabooWord}</TabooWordContainer>
			<TabooBody>{children ? children : <TabooList list={list} />}</TabooBody>
		</StyledTabooCard>
	)
}

TabooCard.propTypes = {
	tabooWord: PropTypes.string.isRequired,
	list: PropTypes.array,
	children: PropTypes.node,
}

//Same as TabooCard, but includes 2 -3 buttons
export const ButtonTabooCard = ({ tabooWord, list, buttons, children, error }) => {
	return (
		<StyledTabooCard>
			<TabooWordContainer>{tabooWord}</TabooWordContainer>
			<TabooBody>
				{children ? children : <TabooList list={list} />}
				{error && <ErrorMessage error={error} />}
			</TabooBody>
			<ButtonContainer buttons={buttons} />
		</StyledTabooCard>
	)
}

ButtonTabooCard.propTypes = {
	tabooWord: PropTypes.string.isRequired,
	buttons: PropTypes.array.isRequired,
	list: PropTypes.array,
	children: PropTypes.node,
}

//Used for background element
export const LayeredTabooCard = ({ tabooWord, list, layer }) => {
	return (
		<StyledLayeredTabooCard layer={layer}>
			<TabooWordContainer>{tabooWord}</TabooWordContainer>
			<TabooBody>
				<TabooList list={list} />
			</TabooBody>
		</StyledLayeredTabooCard>
	)
}

LayeredTabooCard.propTypes = {
	tabooWord: PropTypes.string.isRequired,
	list: PropTypes.array.isRequired,
	layer: PropTypes.string.isRequired,
}

//Used in TabooCard, ButtonTabooCard and FilteredListTabooCard
const TabooList = ({ list, property, specialValue, specialKey, secondary }) => {
	return (
		<List>
			{list.map((listItem, index) => {
				let displayText
				let includeStar = false
				if (typeof listItem === 'object' && listItem !== null && listItem[property]) {
					displayText = listItem[property]

					if (listItem[specialKey] && listItem[specialKey] === specialValue) {
								includeStar = true
					}
				} else {
					displayText = listItem
				}
				return (
					<TabooListItem listItem={displayText} key={index} includeStar={includeStar} secondary={secondary} />
				)
			})}
		</List>
	)
}

TabooList.propTypes = {
	list: PropTypes.array.isRequired,
	property: PropTypes.string,
	specialValue: PropTypes.string,
	specialKey: PropTypes.string,
	secondary: PropTypes.bool,
}

//exported for use in tests
export const TabooListItem = ({ listItem, includeStar, secondary }) => {
	//secondary styling and option to include Star
	return secondary ? (
		<StyledListItemSecondary>
			{listItem}
			{includeStar && <Star />}
		</StyledListItemSecondary>
	) : (
		<StyledListItem>{listItem}</StyledListItem>
	)
}

TabooListItem.propTypes = {
	listItem: PropTypes.any,
	includeStar: PropTypes.bool,
	secondary: PropTypes.bool,
}

export const TabooCardTop = ({ children, margin = false }) => {
	return (
		<StyledTabooCard margin={margin}>
			<TabooWordContainer>{children}</TabooWordContainer>
		</StyledTabooCard>
	)
}

TabooCardTop.propTypes = {
	// children: PropTypes.string.isRequired,
	margin: PropTypes.bool,
}

//ex: playerList array of player objects need to be filtered by team: playerList, team, team1, name, Players, [{}]
export const FilteredTabooList = ({
	unfilteredList,
	listTitle,
	filterKey,
	filterValue,
	displayProperty,
	specialKey,
	specialValue,
	noneMessage,
}) => {

	//filters array based on match of item or item[filterKey] equal to filterValue
	const filteredList = unfilteredList.filter((item) => {
		//If item is an object, filter based on property  value. Otherwise filter based on value of item (i.e. string)
		if (typeof item === 'object' && item !== null) return item[filterKey] === filterValue
		return item === filterValue
	})

	//Displays List of filtered values based on the displayProperty
	return (
		<>
			<StyledListTitle>{listTitle}</StyledListTitle>
			{filteredList.length > 0 ? (
				<TabooList
					list={filteredList}
					property={displayProperty}
					specialKey={specialKey}
					specialValue={specialValue}
					secondary={true}
				/>
			) : (
				<TabooList list={[noneMessage]} secondary={true} />
			)}
		</>
	)
}
