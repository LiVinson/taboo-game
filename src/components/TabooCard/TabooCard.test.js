import React from "react"
import TabooCard from "./TabooCard"
import ButtonGroup from "../ButtonGroup"
import { shallow, mount } from "enzyme"

it("renders div with 'type' class of prop provided ", () => {
  const props = {
    type: "home",
    tabooWord: "taboo",
    list: ["word1", "word2", "word3"],

  }
  const TabooCardComponent = shallow(<TabooCard {...props} />)
  expect(TabooCardComponent.hasClass("taboo-card--home"))
})

it("Renders children prop if it is provided", () => {
  const props = {
    type: "home",
    tabooWord: "taboo",

  }

  const TabooCardComponent = shallow(
    <TabooCard {...props}>
      <h1>I am a child property</h1>
    </TabooCard>
  )
  expect(TabooCardComponent.contains(<h1>I am a child property</h1>)).toEqual(true)
})

it("Renders list prop if children prop is not passed", () => {
    const props = {
      type: "home",
      tabooWord: "taboo",
      list: ["word1", "word2", "word3"],
 
    }
  
    const TabooCardComponent = mount(<TabooCard {...props} />)
    expect(TabooCardComponent.prop("children")).toEqual(undefined)
    
    expect(TabooCardComponent.find("ul.taboo-card__list").children()).toHaveLength(props.list.length)

  })

  it("renders ButtonGroup if button prop is provided", () => {
    const props = {
        type: "home",
        tabooWord: "taboo",
        list: ["word1", "word2", "word3"],
        buttons: [{text: "Back", handleClick: jest.fn()}, {text: "Next", handleClick: jest.fn()}]
      }
      const TabooCardComponent = mount(<TabooCard {...props}/>)
      expect(TabooCardComponent.find(ButtonGroup)).toHaveLength(1)
  })


