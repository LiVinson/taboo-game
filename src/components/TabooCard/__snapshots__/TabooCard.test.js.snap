// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="taboo-card taboo-card--home"
>
  <div
    className="taboo-card__word-container"
  >
    <p
      className="taboo-card__word"
    >
      taboo
    </p>
  </div>
  <div
    className="taboo-card__list-container"
  >
    <ul
      className="taboo-card__list"
    >
      <li
        className="taboo-card__list-item"
      >
        word1
      </li>
      <li
        className="taboo-card__list-item"
      >
        word2
      </li>
      <li
        className="taboo-card__list-item"
      >
        word3
      </li>
    </ul>
  </div>
</div>
`;
